[AArch64][SVE] Optimize logical ops with convert.to.svbool

miloserdow opened the pull request:
When both operands of a logical operation (and/or/xor) are convert.to.svbool
from the same narrower type, unwrap to that type, simplify using simplifyBinOp,
and rewrap the result. This eliminates redundant instructions in cases like:
 svand_z(svptrue_b8(), svpnext_b16(prev, pg), svptrue_b16());

Fixes https://github.com/llvm/llvm-project/issues/160279.

miloserdow committed:
[AArch64][SVE] Optimize logical ops with convert.to.svbool

When both operands of a logical operation (and/or/xor) are convert.to.svbool
from the same narrower type, unwrap to that type, simplify using simplifyBinOp,
and rewrap the result. This eliminates redundant instructions in cases like:
 svand_z(svptrue_b8(), svpnext_b16(prev, pg), svptrue_b16());

Fixes #160279.

miloserdow committed:
[AArch64][SVE] Optimize logical ops with convert.to.svbool

When both operands of a logical operation (and/or/xor) are convert.to.svbool
from the same narrower type, unwrap to that type, simplify using simplifyBinOp,
and rewrap the result. This eliminates redundant instructions in cases like:
 svand_z(svptrue_b8(), svpnext_b16(prev, pg), svptrue_b16());

Fixes #160279.

paulwalker-arm commented:
Rather than adding these explicit combines do you mind trying to extend `constructSVEIntrinsicInfo` and probably `simplifySVEIntrinsicBinOp` to include coverage for the zeroing logical operations? It might not work out but if it does then we'll effectively get the core simplifications for free.

miloserdow commented:
@paulwalker-arm Thanks for the tip! I reworked the simplification to happen through `simplifySVEIntrinsicBinOp`.

miloserdow committed:
Merge branch 'main' into sve-intrinsic-add-or-with-all-true

paulwalker-arm commented:
@miloserdow : Are you still pursuing this work? I recall it's mostly complete with only a few cosmetic requests remaining?

miloserdow commented:
> @miloserdow : Are you still pursuing this work? I recall it's mostly complete with only a few cosmetic requests remaining?

@paulwalker-arm Yes, I completely forgot about this PR for a while. I addressed your last comments and rebased it

paulwalker-arm commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -2014,6 +2014,34 @@ simplifySVEIntrinsicBinOp(InstCombiner &IC, IntrinsicInst &II,
+                     m_Value(NarrowOp2))) &&
+      NarrowOp1->getType() == NarrowOp2->getType() &&
+      NarrowOp1->getType()->isScalableTy() &&
+      NarrowOp1->getType()->isIntOrIntVectorTy(1)) {
```
You can drop these two checks because `aarch64_sve_convert_to_svbool` only support scalable boolean vectors so anything else means the IR is malformed.

paulwalker-arm commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -2014,6 +2014,34 @@ simplifySVEIntrinsicBinOp(InstCombiner &IC, IntrinsicInst &II,
+      else
+        SimpleII = IC.Builder.CreateIntrinsic(
+            Intrinsic::aarch64_sve_convert_to_svbool, {SimpleNarrow->getType()},
+            {SimpleNarrow});
```
I believe you can drop the `{}` wrapping the type and operand and it will still work as expected.
